File: app/src/main/java/com/google/samples/apps/topeka/widget/quiz/MultiSelectQuizView.java
/*
 * Copyright 2015 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.samples.apps.topeka.widget.quiz;

import android.annotation.SuppressLint;
import android.content.Context;
import android.os.Bundle;
import android.util.SparseBooleanArray;
import android.view.View;
import android.widget.AbsListView;
import android.widget.AdapterView;
import android.widget.ListView;

import com.google.samples.apps.topeka.adapter.OptionsQuizAdapter;
import com.google.samples.apps.topeka.helper.AnswerHelper;
import com.google.samples.apps.topeka.model.Category;
import com.google.samples.apps.topeka.model.quiz.MultiSelectQuiz;

@SuppressLint("ViewConstructor")
public class MultiSelectQuizView extends AbsQuizView<MultiSelectQuiz> {

    private static final String KEY_ANSWER = "ANSWER";

    private ListView mListView;

    public MultiSelectQuizView(Context context, Category category, MultiSelectQuiz quiz) {
        super(context, category, quiz);
    }

    @Override
    protected View createQuizContentView() {
        mListView = new ListView(getContext());
        mListView.setAdapter(
                new OptionsQuizAdapter(getQuiz().getOptions(),
                        android.R.layout.simple_list_item_multiple_choice));
        mListView.setChoiceMode(AbsListView.CHOICE_MODE_MULTIPLE);
        mListView.setItemsCanFocus(false);
        mListView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                allowAnswer();
            }
        });
        return mListView;
    }

    @Override
    protected boolean isAnswerCorrect() {
        final SparseBooleanArray checkedItemPositions = mListView.getCheckedItemPositions();
        final int[] answer = getQuiz().getAnswer();
        return AnswerHelper.isAnswerCorrect(checkedItemPositions, answer);
    }

    @Override
    public Bundle getUserInput() {
        Bundle bundle = new Bundle();
        boolean[] bundleableAnswer = getBundleableAnswer();
        bundle.putBooleanArray(KEY_ANSWER, bundleableAnswer);
        return bundle;
    }

    @Override
    public void setUserInput(Bundle savedInput) {
        if (savedInput == null) {
            return;
        }
        final boolean[] answers = savedInput.getBooleanArray(KEY_ANSWER);
        if (null == answers) {
            return;
        }
        for (int i = 0; i < answers.length; i++) {
            mListView.setItemChecked(i, answers[i]);
        }
    }

    private boolean[] getBundleableAnswer() {
        SparseBooleanArray checkedItemPositions = mListView.getCheckedItemPositions();
        final int answerSize = checkedItemPositions.size();
        if (0 == answerSize) {
            return null;
        }
        final int optionsSize = getQuiz().getOptions().length;
        boolean[] bundleableAnswer = new boolean[optionsSize];
        int key;
        for (int i = 0; i < answerSize; i++) {
            key = checkedItemPositions.keyAt(i);
            bundleableAnswer[key] = checkedItemPositions.valueAt(i);
        }
        return bundleableAnswer;
    }
}

File: base/src/main/java/com/google/samples/apps/topeka/widget/quiz/MultiSelectQuizView.kt
/*
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.samples.apps.topeka.widget.quiz

import android.annotation.SuppressLint
import android.content.Context
import android.os.Bundle
import android.view.View
import android.widget.AbsListView
import android.widget.AdapterView
import android.widget.ListView

import com.google.samples.apps.topeka.adapter.OptionsQuizAdapter
import com.google.samples.apps.topeka.helper.AnswerHelper
import com.google.samples.apps.topeka.model.Category
import com.google.samples.apps.topeka.model.quiz.MultiSelectQuiz

@SuppressLint("ViewConstructor")
class MultiSelectQuizView(
        context: Context,
        category: Category,
        quiz: MultiSelectQuiz
) : AbsQuizView<MultiSelectQuiz>(context, category, quiz) {

    private var listView: ListView? = null

    override fun createQuizContentView(): View {
        return ListView(context).apply {
            listView = this
            adapter = OptionsQuizAdapter(quiz.options,
                    android.R.layout.simple_list_item_multiple_choice)
            choiceMode = AbsListView.CHOICE_MODE_MULTIPLE
            itemsCanFocus = false
            onItemClickListener = AdapterView.OnItemClickListener { _, _, _, _ ->
                allowAnswer()
            }
        }
    }

    override val isAnswerCorrect: Boolean
        get() {
            val checkedItemPositions = listView?.checkedItemPositions
            val answer = quiz.answer
            return if (checkedItemPositions != null)
                AnswerHelper.isAnswerCorrect(checkedItemPositions, answer)
            else false
        }

    override var userInput: Bundle
        get() = Bundle().apply { putBooleanArray(KEY_ANSWER, bundleableAnswer) }
        set(savedInput) {
            val answers = savedInput.getBooleanArray(KEY_ANSWER) ?: return
            answers.indices.forEach { listView?.setItemChecked(it, answers[it]) }
        }

    private val bundleableAnswer: BooleanArray?
        get() {
            val checkedItemPositions = listView?.checkedItemPositions ?: return null
            val answerSize = checkedItemPositions.size()
            if (answerSize == 0) {
                return null
            }
            val optionsSize = quiz.options.size
            val bundleableAnswer = BooleanArray(optionsSize)
            (0 until answerSize).forEach {
                bundleableAnswer[checkedItemPositions.keyAt(it)] = checkedItemPositions.valueAt(it)
            }
            return bundleableAnswer
        }
}
